token/cli/src/sort.rs

```diff
@@ -1,99 +1,116 @@
-    program_name == "spl-token" || program_name == "spl-token-2022"
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub(crate) enum AccountFilter {
+    Delegated,
+    Closeable,
+    All,
 }
 
 pub(crate) fn sort_and_parse_token_accounts(
     owner: &Pubkey,
     accounts: Vec<RpcKeyedAccount>,
-    program_id: &Pubkey,
-) -> (MintAccounts, Vec<UnsupportedAccount>, usize, bool) {
-    let mut mint_accounts: MintAccounts = BTreeMap::new();
+    explicit_token: bool,
+    account_filter: AccountFilter,
+) -> Result<CliTokenAccounts, Error> {
+    let mut cli_accounts: BTreeMap<(Pubkey, Pubkey), Vec<CliTokenAccount>> = BTreeMap::new();
     let mut unsupported_accounts = vec![];
     let mut max_len_balance = 0;
     let mut includes_aux = false;
+
     for keyed_account in accounts {
-        let address = keyed_account.pubkey;
+        let address_str = keyed_account.pubkey;
+        let address = Pubkey::from_str(&address_str)?;
+        let program_id = Pubkey::from_str(&keyed_account.account.owner)?;
 
         if let UiAccountData::Json(parsed_account) = keyed_account.account.data {
-            if !is_supported_program(&parsed_account.program) {
-                unsupported_accounts.push(UnsupportedAccount {
-                    address,
-                    err: format!("Unsupported account program: {}", parsed_account.program),
-                });
-            } else {
-                match serde_json::from_value(parsed_account.parsed) {
-                    Ok(TokenAccountType::Account(ui_token_account)) => {
-                        let mint = ui_token_account.mint.clone();
-                        let is_associated = if let Ok(mint) = Pubkey::from_str(&mint) {
-                            spl_associated_token_account::get_associated_token_address_with_program_id(owner, &mint, program_id).to_string() == address
-                        } else {
-                            includes_aux = true;
-                            false
-                        };
-                        let len_balance = ui_token_account
+            match serde_json::from_value(parsed_account.parsed) {
+                Ok(TokenAccountType::Account(ui_token_account)) => {
+                    let mint = Pubkey::from_str(&ui_token_account.mint)?;
+                    let btree_key = (program_id, mint);
+                    let is_associated =
+                        get_associated_token_address_with_program_id(owner, &mint, &program_id)
+                            == address;
+
+                    match account_filter {
+                        AccountFilter::Delegated if ui_token_account.delegate.is_none() => continue,
+                        AccountFilter::Closeable if ui_token_account.close_authority.is_none() => {
+                            continue
+                        }
+                        _ => (),
+                    }
+
+                    if is_associated {
+                        includes_aux = true;
+                    }
```

Comment:
This seems like it should be reversed, no?
```suggestion
                    if !is_associated {
                        includes_aux = true;
                    }
```